lab/skel_graph/astar/main.cpp

```diff
@@ -0,0 +1,228 @@
+
+
+    void read_input() {
+        std::cin >> n >> m;
+        adj_matrix = std::vector<std::vector<int>>(n + 1, std::vector<int> (m + 1 , 0));
+        parent = std::vector<std::vector<Cell>>(
+        n + 1, std::vector<Cell>
+        (m + 1, {-1, -1}));
+        fgh =  std::vector<std::vector<std::tuple<double, double, double>>>
+        (n + 1, std::vector<std::tuple<double, double, double>>
+        (m + 1, std::make_tuple(DBL_MAX, DBL_MAX, DBL_MAX)));
+
+        for (int i = 1; i <= n; i++) {
+            for (int j = 1; j <= m; j++) {
+                std::cin >> adj_matrix[i][j];
+            }
+        }
+
+
+        std::cin >> source.first >> source.second;
+        std::cin >> destination.first >> destination.second;
+    }
+
+
+    void print_output() {
+        for (auto &pos : build_path(destination)) {
+            std::cout<< "Trec prin punctul" << pos.first << " " << pos.second;
+            std::cout << "\n";
+        }
+    }
+
+    std::vector<std::pair <int, int>>  expand(std::pair <int, int> node) {
+        std::vector<std::pair <int, int>> directions =
+        {{1, 1}, {1, 0}, {1, -1}, {0, 1}, {0, -1}, {-1, 1}, {-1, 0}, {-1, -1},};
+        std::vector<std::pair <int, int>> succ;
+        for (const auto &[dx, dy] : directions) {
+            int x = node.first + dx;
+            int y = node.second + dy;
+            if (x >= 1 && x <= n && y >= 1 && y <= n && adj_matrix[x][y] == 1) {
+                succ.push_back({x, y});
+            }
+        }
+        return succ;
+    }
+
+    bool is_goal(const Cell &node) {
+        return (node == destination);
+    }
+
+    double euclidian_distance_heuristic(Cell node) {
+        int abs_x = abs(node.first - destination.first);
+        int abs_y = abs(node.second - destination.second);
+        return sqrt(abs_x * abs_x + abs_y * abs_y);
+    }
+
+    std::vector<std::pair <int, int>> build_path(Cell node) {
+        auto current = node;
+        std::vector<Cell> path;
+        while (!(current.first == -1 && current.second == -1)) {
+            path.push_back(current);
+            current = parent[current.first][current.second];
+        }
+        std::reverse(path.begin(), path.end());
+        return path;
+    }
+
+    bool astar(const std::pair<int, int> &start_state) {
+         // initializari
+        open.clear();  // multimea nodurilor in curs de explorare
+        closed.clear();  // multimea nodurilor expandate
+
+        // functie lambda pentru a compara doua perechi
+        auto comparator = [](auto x1, auto x2) {
+        return x1.first > x2.first;
+        };
+
+        //  priority queue cu nodurile in curs de explorare
+        std::priority_queue<std::pair<double, Cell>,
+        std::vector<std::pair <double, Cell>>,
+        decltype(comparator)> openlist(comparator);
+
+        // pentru s_i avem distanta 0 si parinte NULL
+        std::get<1>(fgh[start_state.first][start_state.second]) = 0.0;
```

Comment:
Please completely remove the tuple. We need an unordered map for each (f, g, h).